bluefog/common/operations.cc

```diff
@@ -713,54 +719,82 @@ void PerformOperation(std::vector<TensorTableEntry>& entries) {
-            request.request_type() != Request::ALLGATHER &&
-            request.request_type() != Request::BROADCAST &&
-            request.request_type() != Request::NEIGHBOR_ALLREDUCE &&
-            request.request_type() != Request::NEIGHBOR_ALLGATHER &&
-            request.request_type() != Request::WIN_CREATE &&
-            request.request_type() != Request::WIN_FREE);
-  };
-  // For these no need to coordinate, put them into entries directly.
-  for (auto& request : message_queue_buffer) {
-    if (IsRequestConvertToEntryDirectly(request)) {
-      entries.push_back(
-          state.tensor_queue.GetTensorEntriesFromRequestDirectly(request));
+  // Wait for all data are ready.
+  for (auto& entry : entries) {
+    if (entry.ready_event != nullptr) {
+      while (!entry.ready_event->Ready()) {
+        std::this_thread::sleep_for(std::chrono::nanoseconds(100));
+      }
     }
   }
-  message_queue_buffer.erase(
-      std::remove_if(message_queue_buffer.begin(), message_queue_buffer.end(),
-                     IsRequestConvertToEntryDirectly),
-      message_queue_buffer.end());
 
-  // For the rest requests, they needs to coordinate and neogiate.
-  // Collect all tensors that are ready to be reduced. Record them in the
-  // tensor count table (rank zero) or send them to rank zero to be
-  // recorded (everyone else).
-  if (global_skip_negotiate_stage) {
-    // Pass don't do anything.
-  } else if (bluefog_rank() == COORDINATE_RANK) {
+  switch (first_entry.mpi_ops_type) {
+    case MPIOpsType::ALLREDUCE:
+      BFLOG(TRACE, bluefog_global.controller->GetRank())
+          << "Processing fused " << first_entry.tensor_name << " and rest "
+          << std::to_string(entries.size()) << " tensors.";
+      timeline.ActivityStartAll(entries, "PROC_ALLREDUCE");
+      if (controller_vendor == Vendor::MPI) {
+        bluefog_global.controller->Allreduce(entries);
+      }
+#if HAVE_NCCL
+      if (controller_vendor == Vendor::NCCL) {
+        bluefog_global.nccl_controller->Allreduce(entries);
+      }
+#endif
+      timeline.ActivityEndAll(entries);
+      break;
+    case MPIOpsType::NEIGHBOR_ALLREDUCE:
+      BFLOG(TRACE, bluefog_global.controller->GetRank())
+          << "Processing fused " << first_entry.tensor_name << " and rest "
+          << std::to_string(entries.size()) << " tensors.";
+      timeline.ActivityStartAll(entries, "PROC_NEIGHBOR_ALLREDUCE");
+      if (controller_vendor == Vendor::MPI) {
+        bluefog_global.controller->NeighborAllreduce(entries);
+      }
+#if HAVE_NCCL
+      if (controller_vendor == Vendor::NCCL) {
+        bluefog_global.nccl_controller->NeighborAllreduce(entries);
+      }
+#endif
+      timeline.ActivityEndAll(entries);
+      break;
+    default:
+      throw std::runtime_error(
+          "Only allreduce or neighbor_allreduce should be called within "
+          "PerformOperationWithFusion");
+  }
+}
+
+void NegotiationOfRequest(BluefogGlobalState& state,
```

Comment:
This function seems very long, consider wrap the fusion step into a separate function.

Reply:
Split it into master and slave two funcions